src/Symfony/Component/Messenger/Transport/Schedule/Schedule.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Schedule;
+
+use Symfony\Component\Clock\ClockInterface;
+
+/**
+ * @internal used to bypass a different evaluation order issue in class constant initializers
+ *
+ * @see https://wiki.php.net/rfc/new_in_initializers#unsupported_positions
+ */
+const _SCHEDULE_FAR_FUTURE_ = new \DateTimeImmutable('3000-01-01');
+
+class Schedule implements \IteratorAggregate
+{
+    public const FAR_FUTURE = _SCHEDULE_FAR_FUTURE_;
+
+    private ClockInterface $clock;
+
+    /**
+     * @var ScheduledMessageInterface[]
+     */
+    private array $scheduledMessages;
+
+    private \DateTimeImmutable $lastTick;
+    private \DateTimeImmutable $nextRun;
+
+    public function __construct(ClockInterface $clock, ScheduledMessageInterface ...$scheduledMessages)
+    {
+        $this->clock = $clock;
+        $this->scheduledMessages = $scheduledMessages;
+
+        $this->lastTick = $this->nextRun = $clock->now();
+    }
+
+    public function getIterator(): \Generator
+    {
+        $currentTime = $this->clock->now();
+        $lastTick = $this->lastTick;
+        $this->lastTick = $currentTime;
+
+        if ($currentTime < $this->nextRun) {
+            return;
+        }
+
+        $nearestNextRun = self::FAR_FUTURE;
+        foreach ($this->scheduledMessages as $scheduledMessage) {
```

Comment:
the implementation is not optimal,
- schedulers are called twice (once for the next run in the future, and once again when lastTick<nextRun<now)
- all schedulers are checked every time one of then is run (if you have one scheduler for `PT1S` and 100 schedulers for `P1D` the 100 schedulers are checked everyv second even if they run once a day)

what's about initializing a table of next runs, then using that table to check if messages should be yield. When a message is yield then add the next run in the table.

Reply:
It's interesting that you've found it :+1:

> schedulers are called twice

On the first run only, I have fixed it.

> all schedulers are checked every time one of then is run

Yes. It significant simplifies an implementation.

> what's about initializing a table of next runs

I have thought about it. I believe it is not necessary. It's a common dilemma: to remember or to compute. If a call `getNextRun()` would be expensive it can be cached (internally), the `PeriodicalTrigger` is simple enough to be called every time.

I'm preparing the next version of implementation. It will work differently and will have different issues :)

Reply:
I have used the `SplHeap` as a "table". Looks good. Could you check?